source/common/http/filter/ratelimit.cc

```diff
@@ -25,16 +25,32 @@ FilterHeadersStatus Filter::decodeHeaders(HeaderMap& headers, bool) {
     cluster_ = config_->cm().get(route->clusterName());
 
     std::vector<::RateLimit::Descriptor> descriptors;
-    for (const Router::RateLimitPolicyEntry& rate_limit :
+
+    // Get all applicable rate limit policy entries for the route.
+    for (const Router::RateLimitPolicyEntry& route_rate_limit :
          route->rateLimitPolicy().getApplicableRateLimit(config_->stage())) {
-      const std::string& route_key = rate_limit.routeKey();
+      const std::string& route_key = route_rate_limit.routeKey();
+      if (!route_key.empty() &&
```

Comment:
You can create a small helper method which takes a list of policies, and then just call it twice (once for route and once for vhost) then you don't duplicate all of this code.